returnn/tf/engine.py

```diff
@@ -2285,15 +2285,18 @@ def search(self, dataset, do_eval=True, output_layer_names="output", output_file
         print("Given output %r has beam %s." % (output_layer, out_beam), file=log.v1)
         output_layer_beam_scores.append(output_layer.get_search_choices().beam_scores)
       out_beam_sizes.append(out_beam.beam_size if out_beam else None)
-      target_keys.append(output_layer.target or self.network.extern_data.default_target)
+      target_key = output_layer.target
+      if output_layer.output.sparse and not target_key:
+        # Use the default target key for sparse outputs
+        target_key = self.network.extern_data.default_target
+      target_keys.append(target_key)
 
     out_cache = None
     seq_idx_to_tag = {}
     if output_file:
       assert output_file_format in {"txt", "py"}
       if output_is_dict:
         assert output_file_format == "py", "Text format not supported in the case of multiple output layers."
-      assert all(dataset.can_serialize_data(target_key) for target_key in target_keys)
```

Comment:
Why is this removed?

Reply:
If this is for the case of `target_key is None`, then just add `if target_key`.

Reply:
Why do we wan't to disallow this? I.e. if the dataset cannot serialise a target key, the corresponding layer output is just directly printed / included in the output file. Not sure though whether there is any case in which this is really the desired behaviour.

Reply:
I think I mostly did this to avoid that the user can maybe have mistake. Now, any such mistake (wrong target key which is not serializable) would not throw an error anymore but silently just alter the behavior.

So, the question is, is this ok or not? Some others should give some opinion on this. (@patrick-wilken  or @JackTemaki maybe?)

Reply:
Another point: in the case of e.g. regression tasks we have dense outputs and targets that are not serialisable by the corresponding dataset. But we still might want to output the result.

Reply:
Well, if we want to support printing dense outputs this check is obsolete.
And the new check in the next line conceptually is also not quite correct. It assumes that if the output layer has a target, it should output sparse data. But any layer can have a target, most notably SoftmaxLayer.
The logic should be: if the output is sparse and the layer has a target and the dataset can serialize this target, do the serialization; else output raw data (which might be dense or sparse).
Then also it would be guaranteed that you get an output for all the keys defined by the user.

Reply:
> The logic should be: if the output is sparse and the layer has a target and the dataset can serialize this target, do the serialization; else output raw data (which might be dense or sparse).

This is the current behaviour.

I removed the check in the next line again to avoid confusion.